### PR TITLE
Added config based tests for tf.data dispatcher and worker servers

### PR DESCRIPTION
**Pull request by kvignesh1420**
This PR extends the test cases for:

- `tf.data.experimental.service.DispatchServer`
- `tf.data.experimental.service.WorkerServer`

and modifies the supported protocol docstring for the above two classes.

The "grpc+local" protocol was not being supported while testing using:

```python
def testStartDispatcherWithProtocolConfig(self):
    config = server_lib.DispatcherConfig(port=5000, protocol="grpc+local")
    dispatcher = server_lib.DispatchServer(config=config, start=False)
    dispatcher.start()
```
The test was erroring out with:

```console
RuntimeError: No credentials factory has been registered for protocol grpc+local. The available types are: [ grpc ]
```

### CODE REVIEW

File: tensorflow/python/data/experimental/service/server_lib.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ class DispatcherConfig(
     port: Specifies the port to bind to. A value of 0 indicates that the server
       may bind to any available port.
     protocol: The protocol to use for communicating with the tf.data service.
-      Acceptable values include `"grpc" and "grpc+local"`.
+      Defaults to the only acceptable value of `"grpc"`.
```
Comment (aaudiber) [inline tensorflow/python/data/experimental/service/server_lib.py:41]: How about removing this sentence entirely? Users can register arbitrary protocols by dynamically linking them into the tensorflow binary, so there could be more acceptable values.

**Review by aaudiber — CHANGES_REQUESTED** [review]
Thanks @kvignesh1420

**Comment by kvignesh1420** [comment]
@aaudiber, thanks for the review. Made the changes based on portpicker for picking a random unused port.

File: tensorflow/python/data/experimental/service/server_lib.py (hunk 1 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/service/server_lib.py]: just mentioned it as "Defaults to grpc".

**Comment by gbaned** [comment]
@kvignesh1420 Can you please resolve conflicts? Thanks!

File: tensorflow/python/data/experimental/service/server_lib_test.py [hunk 2]
```diff
@@ -18,8 +18,9 @@
 from __future__ import division
 from __future__ import print_function
 
+import tempfile
+import portpicker
```
Comment (aaudiber) [inline tensorflow/python/data/experimental/service/server_lib_test.py:22]: https://github.com/tensorflow/tensorflow/blob/f74cc7a696c66db173b321d51fb05f032652a6c7/tensorflow/python/distribute/multi_worker_test_base.py#L34 has logic for handling `portpicker` not being available, so I think we'll need to do the same thing here.

**Review by aaudiber — CHANGES_REQUESTED** [review]

File: tensorflow/python/data/experimental/service/server_lib_test.py (hunk 2 above)
Comment (kvignesh1420) [reply tensorflow/python/data/experimental/service/server_lib_test.py]: addressed in the commit

**Comment by kvignesh1420** [comment]
@aaudiber I was a bit dicey about using `threading.Lock()` in `pick_unused_port()` function but included it anyway for additional thread safety while running the tests. Let me know if you feel otherwise.

**Review by aaudiber — APPROVED** [review]

File: tensorflow/python/data/experimental/service/server_lib.py [hunk 3]
```diff
@@ -39,7 +39,7 @@ class DispatcherConfig(
     port: Specifies the port to bind to. A value of 0 indicates that the server
       may bind to any available port.
     protocol: The protocol to use for communicating with the tf.data service.
-      Acceptable values include `"grpc" and "grpc+local"`.
+      Defaults to `"grpc"`.
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/service/server_lib.py:42]: @aaudiber Were the changes to docstrings intentional?

Comment (kvignesh1420) [reply tensorflow/python/data/experimental/service/server_lib.py:42]: @jsimsa 

The "grpc+local" protocol was not being supported while testing using:

```python
def testStartDispatcherWithProtocolConfig(self):
    config = server_lib.DispatcherConfig(port=5000, protocol="grpc+local")
    dispatcher = server_lib.DispatchServer(config=config, start=False)
    dispatcher.start()
```
The test was erroring out with:

```console
RuntimeError: No credentials factory has been registered for protocol grpc+local. The available types are: [ grpc ]
```

Comment (aaudiber) [reply tensorflow/python/data/experimental/service/server_lib.py:42]: Yes, I originally planned to link the grpc+local protocol into the TF binary, but encountered issues using the protocol on windows and decided not to link it by default.

**Pull request merged**